File: lc_soln/0238_products_of_array_except_self/solution.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        fwd = [1] * (n + 2)
        bwd = [1] * (n + 2)
        for i in range(n):
            fwd[i + 1] = fwd[i] * nums[i]
            bwd[n - i] = bwd[n - i + 1] * nums[n - 1 - i]
        result = []
        for i in range(n):
            result.append(fwd[i] * bwd[i + 2])

        return result


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0238_products_of_array_except_self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        std::vector<int> fwd(n);
        std::vector<int> bwd(n);
        fwd[0] = 1;
        bwd[n - 1] = 1;
        for (int i = 1; i < nums.size(); ++i) {
            fwd[i] = fwd[i - 1] * nums[i - 1];
            bwd[n - 1 - i] = bwd[n - i] * nums[n - i];
        }
        std::vector<int> result(n);
        for (int i = 0; i < nums.size(); ++i) {
            result[i] = fwd[i] * bwd[i];
        }
        return result;
    }
};

int main() {
    return 0;
}
